From 91ab26b5e994114cef4e4d162cce698af5e6723a Mon Sep 17 00:00:00 2001
From: He Junyan <junyan.he@intel.com>
Date: Fri, 23 Jun 2023 20:01:04 +0800
Subject: [PATCH 125/125] va: basedec: Select the best format of the whole caps

The current way only selects the best video format from the first
structure of the caps. The caps like:

  video/x-raw(memory:VAMemory),drm-format=(string)NV12;  \
  video/x-raw(memory:VAMemory),format=(string){ NV12, Y210 }

Will just choose NV12 as the result, even the bitstream is 10 bits.
---
 .../gst-plugins-bad/sys/va/gstvabasedec.c     | 23 +++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/subprojects/gst-plugins-bad/sys/va/gstvabasedec.c b/subprojects/gst-plugins-bad/sys/va/gstvabasedec.c
index 42bdcc622b..55efdcf8e2 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvabasedec.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvabasedec.c
@@ -821,13 +821,15 @@ _find_video_format_from_chroma (const GValue * formats, guint chroma_type,
 
 static GstVideoFormat
 _caps_video_format_from_chroma (GstCaps * caps, GstCapsFeatures * features,
-    guint chroma_type, guint64 * modifier)
+    guint chroma_type, guint64 * ret_modifier)
 {
   guint i, num_structures;
   gboolean drm_format;
   GstCapsFeatures *feats;
   GstStructure *structure;
   const GValue *format;
+  GstVideoFormat fmt, ret_fmt = GST_VIDEO_FORMAT_UNKNOWN;
+  guint64 modifier;
 
   num_structures = gst_caps_get_size (caps);
   for (i = 0; i < num_structures; i++) {
@@ -844,11 +846,24 @@ _caps_video_format_from_chroma (GstCaps * caps, GstCapsFeatures * features,
       drm_format = FALSE;
     }
 
-    return _find_video_format_from_chroma (format, chroma_type, drm_format,
-        modifier);
+    fmt = _find_video_format_from_chroma (format, chroma_type, drm_format,
+        &modifier);
+    if (fmt == GST_VIDEO_FORMAT_UNKNOWN)
+      continue;
+
+    if (ret_fmt == GST_VIDEO_FORMAT_UNKNOWN) {
+      ret_fmt = fmt;
+      *ret_modifier = modifier;
+    }
+
+    if (gst_va_chroma_from_video_format (fmt) == chroma_type) {
+      ret_fmt = fmt;
+      *ret_modifier = modifier;
+      break;
+    }
   }
 
-  return GST_VIDEO_FORMAT_UNKNOWN;
+  return ret_fmt;
 }
 
 static GstVideoFormat
-- 
2.25.1

